syntax = "proto3";

package raft.v1;

service RaftService {
     rpc GetLeader(GetLeaderRequest) returns (GetLeaderResponse);
     rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
     rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
     rpc RemovePeer(RemovePeerRequest) returns (RemovePeerResponse);
}

message GetLeaderRequest{}

message GetLeaderResponse {
  string leader_address = 1;
}

message RequestVoteRequest {
  int64 term = 1;
  int64 candidate_id = 2;
  int64 last_log_index = 3;
  int64 last_log_term = 4;
}

message RequestVoteResponse {
  int64 term = 1;
  bool vote_granted = 2;
}

message LogEntry{
  int64 term = 1;
  bytes data = 2;
}

message AppendEntriesRequest {
  int64 term = 1;
  int64 leader_id = 2;
  int64 prev_log_index = 3;
  int64 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int64 leader_commit = 6;
}

message AppendEntriesResponse{
  int64 term = 1;
  bool success = 2;
  int64 match_index = 3;
}

message RemovePeerRequest{
    int64 id = 1;
}

message RemovePeerResponse{
    int64 term = 1;
    bool success = 2;
}